### PR TITLE
More ValueTask goodness for System.IO.Pipelines

### PR DESCRIPTION
**Pull request by davidfowl**
- Use a synchronusly completing ValueTask if the result is completed at the time the async operation is executed.
- Added more tests around ConfigureAwait

### CODE REVIEW

File: src/System.IO.Pipelines/tests/SchedulerFacts.cs [hunk 1]
```diff
@@ -291,11 +397,8 @@ private void Work(object state)
 
             async Task DoRead()
             {
-                int oid = Thread.CurrentThread.ManagedThreadId;
```
Comment (pakrym) [inline src/System.IO.Pipelines/tests/SchedulerFacts.cs]: Why removed?

Comment (davidfowl) [reply src/System.IO.Pipelines/tests/SchedulerFacts.cs]: The thread could be reused. I saw it fail once in a local run.

Comment (pakrym) [reply src/System.IO.Pipelines/tests/SchedulerFacts.cs]: This way test tests nothing, if test starts on threadpool thread and scheduler just runs inline it won't fail

Comment (davidfowl) [reply src/System.IO.Pipelines/tests/SchedulerFacts.cs]: What about this, I'll only assert that thread isn't a thread pool thread to begin with, if it is we'll know why it's failing if it does fail randomly. I'll also add a check to make sure it doesn't complete synchronously.

Comment (pakrym) [reply src/System.IO.Pipelines/tests/SchedulerFacts.cs]: > I'll assert that thread isn't a thread pool thread to begin with

Then keep the other asserts too, they wouldn't fail in other case.

Comment (davidfowl) [reply src/System.IO.Pipelines/tests/SchedulerFacts.cs]: Got it! Confirmed that we do start on the thread pool sometimes. I'll make a manual thread as you suggest.

Comment (davidfowl) [reply src/System.IO.Pipelines/tests/SchedulerFacts.cs]: @pakrym Yea I added them all back and added more as well.

**Comment by pakrym** [comment]
Do we have test where we set useSyncContext:false and set current context checking that it is ignored?

**Comment by davidfowl** [comment]
> Do we have test where we set useSyncContext:false and set current context checking that it is ignored?

Not an explicit one, I'll add one. Since xunit runs with a sync context, every test is testing that. This will be one we'll have to deal with when we take this change into ASP.NET. We may have to turn this off everywhere as well.

**Comment by pakrym** [comment]
Can we ensure somehow that sync valuetask is returned where we expect it?

**Comment by pakrym** [comment]
Do we ever test  FlowExecutionContext branch of PipeAwaitable.OnCompleted? async state machine would call OnUnsafeCompleted that doesn't pass it.

**Comment by davidfowl** [comment]
There’s a test that explicitly tests this code path calling OnCompleted directly in PipeReaderWriterFacts. That’s the only case where it would happen (or if you wrote your own async method builder that only calls OnCompleted on the awaiters it got back)

**Pull request merged**